fggs/multi.py

```diff
@@ -175,6 +175,53 @@ def multi_mv(a: MultiTensor, b: MultiTensor, transpose: bool = False) -> MultiTe
                 c.add_single(x, axy.mv(by, semiring).reshape(ishapes[x]))
     return c
 
+def _order_nonterminals(a: MultiTensor) -> list:
+    """ Reorders the nonterminals in a Multitensor.
+        Uses the heuristic found in Nederhof and Satta (2008) to put the 
+        linking nonterminals first, speeding up Gaussian Elimination
+    """    
+    if not a.keys():
+        return []
+    
+    #set up graph
+    nonterminal_graph: Dict[MultiTensorKey, set] = {}  
+    for x, y in a.keys():
+        if x not in nonterminal_graph:
+            nonterminal_graph[x] = set()
+            
+        nonterminal_graph[x].add(y)
+    
+    #dfs on graph to get finish times
+    finish_times = {}
+    visited = set()
+
+    def dfs(node, iteration):
+        visited.add(node)
+        finish_times[node] = iteration
+        
+        for neighbor in nonterminal_graph.get(node, []):
+            if neighbor not in visited:
+                dfs(neighbor, iteration + 1)
```

Comment:
You're computing the depth in the DFS tree, but the finishing times are supposed to be unique (if u != v then finish[u] != finish[v]). What you did might also work, but I'll have to think about it.

Reply:
![IMG_5896](https://github.com/user-attachments/assets/417bc6e9-a75e-4aa8-95c5-d0b46a90c36d)

Reply:
Assuming that the DFS iterates over children left-to-right (so, A before D), I believe the depths and finishing times above are correct. And according to the current code, I think C and D would both be non-linking nonterminals. But according to the paper, just D would be a non-linking nonterminal.

Reply:
You're definitely right here. I'll make this change.